tools/language-server/logstat/src/main.rs

```diff
@@ -0,0 +1,317 @@
+    max:  Duration,
+    avg:  Duration,
+    line: String,
+}
+
+impl fmt::Display for Stats {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{avg}ms [{min}..{max}] {line}",
+            avg = self.avg.whole_milliseconds(),
+            min = self.min.whole_milliseconds(),
+            max = self.max.whole_milliseconds(),
+            line = self.line.chars().take(80).collect::<String>()
+        )
+    }
+}
+
+lazy_static! {
+    static ref RE_LOG: Regex = Regex::new(r"\[([\w]+)\] \[([\w\d:.-]+)\] (.*)").unwrap();
+    static ref WSTEST_SPEC: Spec = Spec {
+        matches: vec![
+            "wstest sent bench request".to_string(),
+            "wstest handled response".to_string(),
+        ],
+    };
+}
+
+/// Read the file with specification
+async fn read_specs(path: &PathBuf) -> Result<Spec> {
+    let file = File::open(path).await?;
+    let mut lines = BufReader::new(file).lines();
+    let mut matches = Vec::new();
+
+    while let Some(line) = lines.next_line().await? {
+        matches.push(line);
+    }
+
+    Ok(Spec { matches })
+}
+
+/// Extract the iterations information from logfile according to the provided spec
+async fn read_logfile(path: &PathBuf, spec: &Spec) -> Result<Vec<Iteration>> {
+    let file = File::open(path).await?;
+    let mut lines = BufReader::new(file).lines();
+
+    let mut iterations = vec![];
+    let mut current_operations = vec![];
+    let mut matches = spec.matches.iter();
+
+    let mut current_match = if let Some(m) = matches.next() {
+        m
+    } else {
+        matches = spec.matches.iter();
+        matches.next().expect("Empty spec!")
+    };
+
+    while let Some(line) = lines.next_line().await? {
+        if let Some(cap) = RE_LOG.captures(line.as_str()) {
+            let groups = (cap.get(2), cap.get(3));
+            match groups {
+                (Some(timestamp), Some(message)) =>
+                    if message.as_str().contains(current_match) {
+                        let duration = Duration::ZERO;
+                        let timestamp =
+                            OffsetDateTime::parse(timestamp.as_str(), &Rfc3339).unwrap();
+                        let line = message.as_str().to_string();
+
+                        current_operations.push(Operation { duration, timestamp, line });
+
+                        if let Some(m) = matches.next() {
+                            current_match = m
+                        } else {
+                            matches = spec.matches.iter();
+                            current_match = matches.next().expect("Empty spec!");
+                            iterations.push(Iteration { operations: current_operations });
+                            current_operations = vec![];
+                        }
+                    },
+                _ => {
+                    eprintln!("[ERR] Invalid log line [{}]", line);
+                }
+            }
+        }
+    }
```

Comment:
This can be written in a clearer way: again, by turning lines into `Stream`, and using [filter_map](https://docs.rs/tokio-stream/0.1.9/tokio_stream/trait.StreamExt.html#method.filter_map). The main branch of match can be moved to a separate function.